src/Cassandra/HostConnectionPool.cs

```diff
@@ -288,13 +308,16 @@ public static Connection MinInFlight(Connection[] connections, ref int connectio
                 //Overflow protection, not exactly thread-safe but we can live with it
                 Interlocked.Exchange(ref connectionIndex, 0);
             }
+
             Connection c = null;
+            inFlight = 0;
+
             for (var i = index; i < index + connections.Length; i++)
             {
                 c = connections[i % connections.Length];
                 var previousConnection = connections[(i - 1) % connections.Length];
```

Comment:
A potential simplification. It also has one less modulo per iteration. This breaks the semantics/ordering a bit (but it matches the description `index + 1 and index` ;p ) so the unit tests will need to be updated.

I moved the invariant `i + index` into the loop instead of being in the for loop setup. I find loops like this to be easier to understand and the compiler will almost certainly detect this type of invariant. 

```cs
            Connection c = connections[index % connections.Length];
            inFlight = c.InFlight;

            for (var i = 1; i < connections.Length; i++) // Start at the next index
            {
                var nextConnection = connections[(i + index) % connections.Length];
                var nextInflight = nextConnection.InFlight; // Avoid multiple volatile reads
                if (nextInflight < inFlight)
                {
                    inFlight = nextInflight;
                    c = nextConnection;
                }
                if (inFlight < inFlightThreshold)
                {
                    // We should avoid traversing all the connections
                    // We have a connection with a decent amount of in-flight requests
                    break;
                }
            }
```

Reply:
> This could result in a negative index? Seems weird.

`Interlocked.Increment()` returns the result, so it's going to be 1 or greater.

In any case, your suggestion for the loop using `var i = 1; i < connections.Length; i++` makes a lot of sense, its far more readable, I'll change it.

Reply:
Not in the case of an overflow. Unlikely to happen but still possible?

Reply:
Make the other code post increment :)

```cs
int index, nextIndex;
do
{
    index = connectionIndex;
    nextIndex = index + 1;
    if (nextIndex > ConnectionIndexOverflow)
    {
        nextIndex = 0;
    }
} while (Interlocked.CompareExchange(ref connectionIndex, nextIndex, index) != index);
```

Reply:
Or you could do `Interlocked.Increment() - 1`.

Reply:
In the case of reaching the overflow threshold, the `field` containing the previous positive value is still used (a value greater than `ConnectionIndexOverflow`), while the shared `connectionIndex` is set to `0`. The following time, its going to be incremented again and it's going to be > 0. As far as I can see it, there is no risk of `(i - 1)` being negative.

Reply:
yeah. That makes sense. I see `i - 1` as an array index and it makes alarms go off in my head.

Reply:
Indeed! I'll change it to your suggestion.